geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/collections/SizeableByteArrayList.java

```diff
@@ -0,0 +1,130 @@
+
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.ListIterator;
+
+import org.apache.geode.internal.size.Sizeable;
+
+public class SizeableByteArrayList extends LinkedList<byte[]> implements Sizeable {
+  private static final int BYTE_ARRAY_LIST_OVERHEAD = memoryOverhead(SizeableByteArrayList.class);
+  private static final int NODE_OVERHEAD =
+      roundUpSize(getObjectHeaderSize() + 3 * getReferenceSize());
+  private static final int BYTE_ARRAY_BASE_OVERHEAD = 16;
+  private int memberOverhead;
+
+  @Override
+  public int indexOf(Object o) {
+    ListIterator<byte[]> iterator = this.listIterator();
+    while (iterator.hasNext()) {
+      int index = iterator.nextIndex();
+      byte[] element = iterator.next();
+      if (Arrays.equals(element, (byte[]) o)) {
+        return index;
+      }
+    }
+    return -1;
+  }
+
+  @Override
+  public int lastIndexOf(Object o) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public boolean remove(Object o) {
+    ListIterator<byte[]> iterator = this.listIterator();
+    while (iterator.hasNext()) {
+      byte[] element = iterator.next();
+      if (Arrays.equals(element, (byte[]) o)) {
+        iterator.remove();
+        memberOverhead -= calculateByteArrayOverhead(element);
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public byte[] remove(int index) {
+    byte[] element = super.remove(index);
+    memberOverhead -= calculateByteArrayOverhead(element);
+    return element;
+  }
+
+  @Override
+  public void addFirst(byte[] element) {
+    memberOverhead += calculateByteArrayOverhead(element);
+    super.addFirst(element);
+  }
+
+  @Override
+  public void addLast(byte[] element) {
+    memberOverhead += calculateByteArrayOverhead(element);
+    super.addLast(element);
+  }
+
+  public boolean removeLastOccurrence(Object o) {
+    throw new UnsupportedOperationException();
+  }
+
+  private int calculateByteArrayOverhead(byte[] element) {
+    return BYTE_ARRAY_BASE_OVERHEAD + (element.length % 8 == 0 ? 0 : 8) +
+        NODE_OVERHEAD + (element.length / 8) * 8;
+  }
+
+  @Override
+  public int getSizeInBytes() {
+    return BYTE_ARRAY_LIST_OVERHEAD + memberOverhead;
+  }
+
+  @Override
+  public int hashCode() {
+    final int primeNumber = 31;
+    int hashCode = 1;
+    ListIterator<byte[]> iterator = this.listIterator();
+    while (iterator.hasNext()) {
+      hashCode = hashCode * primeNumber + Arrays.hashCode(iterator.next());
+    }
+    return hashCode;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof SizeableByteArrayList)) {
+      return false;
+    }
+    SizeableByteArrayList sizeableByteArrayList = (SizeableByteArrayList) o;
+    if (sizeableByteArrayList.size() != this.size()) {
+      return false;
+    }
+    for (int i = 0; i < sizeableByteArrayList.size(); i++) {
```

Comment:
given that these are linked lists it seems wrong keep calling get(i) which is going to need to scan the list each time because random access is not supported. I think it would be better to create two iterators, one for each list, and then loop around advancing each iterator on each iteration.

Reply:
O(N^2) to O(N).